### PR TITLE
log: Introduce EnabledParameters

### PR DESCRIPTION
**Pull request by pellared**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/5769

Related spec PR: https://github.com/open-telemetry/opentelemetry-specification/pull/4203

Remark: A follow-up in contrib is required and afterwards here to fix the `example/dice`.

Benchstat results for `sdk/log` (`log` has no benchmarks related to Enabled):

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/log
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
                                  │    old.txt    │               new.txt                │
                                  │    sec/op     │    sec/op     vs base                │
BatchProcessorOnEmit-16              398.8n ± 10%   395.6n ± 12%        ~ (p=0.971 n=10)
Processor/Simple-16                  882.2n ±  8%   869.8n ±  9%        ~ (p=0.811 n=10)
Processor/Batch-16                   1.478µ ±  3%   1.485µ ±  5%        ~ (p=0.646 n=10)
Processor/SetTimestampSimple-16      847.8n ±  1%   844.6n ±  3%        ~ (p=0.247 n=10)
Processor/SetTimestampBatch-16       1.480µ ±  3%   1.473µ ±  4%        ~ (p=0.700 n=10)
Processor/AddAttributesSimple-16     930.0n ±  1%   933.8n ±  1%        ~ (p=0.172 n=10)
Processor/AddAttributesBatch-16      1.624µ ±  2%   1.639µ ±  2%        ~ (p=0.839 n=10)
Processor/SetAttributesSimple-16     903.4n ±  1%   895.1n ±  1%        ~ (p=0.190 n=10)
Processor/SetAttributesBatch-16      1.554µ ±  4%   1.529µ ±  3%        ~ (p=0.159 n=10)
LoggerNewRecord/5_attributes-16      346.0n ±  2%   343.3n ±  2%        ~ (p=0.448 n=10)
LoggerNewRecord/10_attributes-16     1.608µ ±  6%   1.503µ ±  2%   -6.53% (p=0.007 n=10)
LoggerEnabled-16                    34.305n ±  8%   6.706n ±  1%  -80.45% (p=0.000 n=10)
LoggerProviderLogger-16              636.9n ± 10%   605.8n ±  3%        ~ (p=0.105 n=10)
WalkAttributes/1_attributes-16       5.363n ±  3%   4.540n ± 14%  -15.34% (p=0.002 n=10)
WalkAttributes/10_attributes-16      5.436n ±  7%   4.461n ±  2%  -17.95% (p=0.000 n=10)
WalkAttributes/100_attributes-16     5.126n ±  9%   4.465n ±  1%  -12.90% (p=0.000 n=10)
WalkAttributes/1000_attributes-16    5.316n ±  9%   4.502n ±  5%  -15.32% (p=0.002 n=10)
SetAddAttributes/SetAttributes-16    220.5n ± 18%   192.6n ± 11%  -12.67% (p=0.007 n=10)
SetAddAttributes/AddAttributes-16    165.3n ± 21%   127.3n ± 22%  -22.96% (p=0.011 n=10)
SimpleProcessorOnEmit-16             2.159n ±  9%   2.167n ±  9%        ~ (p=0.739 n=10)
geomean                              178.3n         154.5n        -13.31%

                        │    old.txt    │             new.txt             │
                        │      B/s      │      B/s       vs base          │
BatchProcessorOnEmit-16   76.52Mi ± 11%   77.14Mi ± 14%  ~ (p=0.971 n=10)

                                  │    old.txt     │                new.txt                │
                                  │      B/op      │     B/op      vs base                 │
BatchProcessorOnEmit-16               0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/Simple-16                   417.0 ± 0%       417.0 ± 0%       ~ (p=1.000 n=10) ¹
Processor/Batch-16                  1.093Ki ± 1%     1.088Ki ± 1%       ~ (p=0.254 n=10)
Processor/SetTimestampSimple-16       417.0 ± 0%       417.0 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetTimestampBatch-16      1.095Ki ± 1%     1.084Ki ± 2%       ~ (p=0.361 n=10)
Processor/AddAttributesSimple-16      417.0 ± 0%       417.0 ± 0%       ~ (p=1.000 n=10) ¹
Processor/AddAttributesBatch-16     1.085Ki ± 1%     1.086Ki ± 1%       ~ (p=1.000 n=10)
Processor/SetAttributesSimple-16      465.0 ± 0%       465.0 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetAttributesBatch-16     1.129Ki ± 1%     1.125Ki ± 1%       ~ (p=0.084 n=10)
LoggerNewRecord/5_attributes-16       0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerNewRecord/10_attributes-16      610.0 ± 0%       610.0 ± 0%       ~ (p=1.000 n=10) ¹
LoggerEnabled-16                      0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerProviderLogger-16               359.0 ± 6%       346.0 ± 3%       ~ (p=0.117 n=10)
WalkAttributes/1_attributes-16        0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/10_attributes-16       0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/100_attributes-16      0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/1000_attributes-16     0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
SetAddAttributes/SetAttributes-16     48.00 ± 0%       48.00 ± 0%       ~ (p=1.000 n=10) ¹
SetAddAttributes/AddAttributes-16     0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
SimpleProcessorOnEmit-16              0.000 ± 0%       0.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                          ²                 -0.27%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                                  │   old.txt    │               new.txt               │
                                  │  allocs/op   │ allocs/op   vs base                 │
BatchProcessorOnEmit-16             0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/Simple-16                 1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/Batch-16                  1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetTimestampSimple-16     1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetTimestampBatch-16      1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/AddAttributesSimple-16    1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/AddAttributesBatch-16     1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetAttributesSimple-16    2.000 ± 0%     2.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetAttributesBatch-16     2.000 ± 0%     2.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerNewRecord/5_attributes-16     0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerNewRecord/10_attributes-16    4.000 ± 0%     4.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerEnabled-16                    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
LoggerProviderLogger-16             1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/1_attributes-16      0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/10_attributes-16     0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/100_attributes-16    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
WalkAttributes/1000_attributes-16   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
SetAddAttributes/SetAttributes-16   1.000 ± 0%     1.000 ± 0%       ~ (p=1.000 n=10) ¹
SetAddAttributes/AddAttributes-16   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
SimpleProcessorOnEmit-16            0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                        ²               +0.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

### CODE REVIEW

File: log/enabled.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package log // import "go.opentelemetry.io/otel/log"
+
+// EnabledParam represents payload for [Logger]'s Enabled method.
+type EnabledParam struct {
```
Comment (dashpole) [inline log/enabled.go:7]: non-blocking: personal preference for `EnabledOpts`.   It is just a naming pattern i've seen in other libraries (e.g. prometheus)

**Review by dashpole — APPROVED** [review]

File: log/enabled.go (hunk 1 above)
Comment (pellared) [reply log/enabled.go:7]: I like the proposal. I had big trouble finding a good name.

Comment (pellared) [reply log/enabled.go]: Renamed https://github.com/open-telemetry/opentelemetry-go/pull/5791/commits/450fc7a0c7eb084f8f7b39ee0a3b194f7ee254c7

File: log/enabled.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package log // import "go.opentelemetry.io/otel/log"
+
+// EnabledOpts represents payload for [Logger]'s Enabled method.
+type EnabledOpts struct {
```
Comment (MrAlias) [inline log/enabled.go:7]: This does not follow our options pattern policy: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#option

Please update to match or use a different name here to distinguish the alternate intent.

Comment (dashpole) [reply log/enabled.go:7]: Which part of the policy does it violate?  Is your main concern user confusion?  Or that it would introduce a new "pattern" which others would want to adopt?

Another idea would be to choose something smilar to Record (since this is details about the log record): RecordDetails, RecordSummary, RecordXXX?  Or we could go back to params

Comment (MrAlias) [reply log/enabled.go:7]: We have a definition for what an "option" is. We need to follow closer to what we do for [`SamplingParameters`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#SamplingParameters), including naming.

Comment (pellared) [reply log/enabled.go:7]: Should I revert to EnabledParam or change to EnabledParameters?

Comment (MrAlias) [reply log/enabled.go:7]: +1 to `EnabledParameters` to match the tracing signal. But, I'm not opposed to `EnabledParam`.

Comment (MrAlias) [reply log/enabled.go:7]: Given this isn't a 1-to-1 matching with the trace SDK's `Sampler` interfaced I think we have some leeway in naming. Just, we need to not conflict with our options naming or pattern.

Comment (pellared) [reply log/enabled.go:7]: I prefer shorter `EnabledParam`, but I am also not opposed to `EnabledParameters`.

I propose a vote for probing.

- 🚀 for `EnabledParam`
- 🎉 for `EnabledParameters`

Comment (pellared) [reply log/enabled.go]: Renamed to `EnabledParameters`. b9a617178

**Review by dmathieu — APPROVED** [review]

File: example/dice/rolldice.go [hunk 3]
```diff
@@ -4,7 +4,10 @@
 package main
 
 import (
-	"go.opentelemetry.io/contrib/bridges/otelslog"
+	// TODO: "go.opentelemetry.io/contrib/bridges/otelslog".
```
Comment (dashpole) [inline example/dice/rolldice.go:7]: I'd rather not leave these TODOs in the example. Maybe use a tracking issue instead?

**Review by dashpole — APPROVED** [review]

File: example/dice/rolldice.go (hunk 3 above)
Comment (pellared) [reply example/dice/rolldice.go:7]: https://github.com/open-telemetry/opentelemetry-go/issues/5801

Comment (pellared) [reply example/dice/rolldice.go]: Also https://github.com/open-telemetry/opentelemetry-go/pull/5791/commits/ae37bbad440ff8403c24fd6ff112801fe4450ba8

File: sdk/log/internal/x/x.go [hunk 4]
```diff
@@ -10,37 +10,38 @@ import (
...
 	// The returned value will be true when the Processor will process for the
-	// provided context and record, and will be false if the Processor will not
+	// provided context and param, and will be false if the Processor will not
 	// process. An implementation should default to returning true for an
 	// indeterminate state.
 	//
-	// Implementations should not modify the record.
-	Enabled(ctx context.Context, record log.Record) bool
+	// Implementations should not modify the param.
+	Enabled(ctx context.Context, param log.EnabledParameters) bool
```
Comment (pellared) [inline sdk/log/internal/x/x.go:46]: Separate issue.

I think that on the SDK level there should be a separate `EnabledParameters` type that includes more information and have getters for `resource.Resource`, `instrumentation.Scope`, `trace.TraceID`, `trace.SpanID`, `trace.TraceFlags`. Similarly like we have a separate `Record` type for SDK which includes this information.

File: example/dice/rolldice.go [hunk 5]
```diff
@@ -14,7 +18,7 @@ const name = "go.opentelemetry.io/otel/example/dice"
 var (
 	tracer  = otel.Tracer(name)
 	meter   = otel.Meter(name)
-	logger  = otelslog.NewLogger(name)
+	logger  = slog.New(slog.NewJSONHandler(os.Stdout, nil)) // TODO: logger  = otelslog.NewLogger(name).
```
Comment (MrAlias) [inline example/dice/rolldice.go:21]: Why is this change being made here? This seems unrelated to introducing the `EnabledParameters` type?

File: log/enabled.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+// Severity returns the [Severity] level.
+func (r *EnabledParameters) Severity() Severity {
+	return r.severity
+}
+
+// SetSeverity sets the [Severity] level.
+func (r *EnabledParameters) SetSeverity(level Severity) {
+	r.severity = level
+}
```
Comment (MrAlias) [inline log/enabled.go:19]: Why have getters and setters here?

```go
type EnabledParameters struct {
	Severity Severity
}
```

This means a user would be able to do `logger.Enabled(ctx, EnabledParameters{Severity: SeverityInfo})` instead of:

```go
p := new(EnabledParameters)
p.SetSeverity(SeverityInfo)
logger.Enabled(ctx, p)
```

Comment (MrAlias) [reply log/enabled.go:19]: Additionally using fields directly matches the `SamplingParameters` this is modeled after.

File: log/enabled.go
Comment (MrAlias) [inline log/enabled.go:1]: Can we move this to where it is used in logger.go?

File: log/enabled.go (hunk 6 above)
Comment (MrAlias) [inline log/enabled.go:19]:
```suggestion
// EnabledParameters represents payload for [Logger]'s Enabled method.
type EnabledParameters struct {
	severity Severity
	severitySet bool
}

// Severity returns the [Severity] level.
func (r *EnabledParameters) Severity() (Severity, bool) {
	return r.severity, r.severitySet
}

// SetSeverity sets the [Severity] level.
func (r *EnabledParameters) SetSeverity(level Severity) {
	r.severity = level
	r.severitySet = true
}
```

File: example/dice/rolldice.go (hunk 5 above)
Comment (pellared) [reply example/dice/rolldice.go:21]: Usage of `otelslog` does not compile in current form because it used `Logger.Enabled` from before the breaking change https://github.com/open-telemetry/opentelemetry-go-contrib/blob/d312469f21ca6606e872292bdca206b2680608d3/bridges/otelslog/handler.go#L207

File: log/enabled.go
Comment (pellared) [reply log/enabled.go:1]: Sure

File: log/enabled.go (hunk 6 above)
Comment (pellared) [reply log/enabled.go:19]: I was thinking about the same 👍

File: example/dice/rolldice.go (hunk 5 above)
Comment (pellared) [reply example/dice/rolldice.go:21]: This is related to https://github.com/open-telemetry/opentelemetry-go/issues/5801. I can add another TODO comment above if you find it necessary.

File: log/enabled.go (hunk 6 above)
Comment (MrAlias) [reply log/enabled.go:19]: Yeah, I think if we are going to go the getter/setter route, we should leverage this type of interface.

Comment (pellared) [reply log/enabled.go]: Addressed https://github.com/open-telemetry/opentelemetry-go/pull/5791/commits/9db67acf75c934899a286d017b8b44b538013679

Doc comments inspired by https://pkg.go.dev/sync#Map.Load

**Comment by pellared** [comment]
CC @bogdandrutu

**Review by MrAlias — APPROVED** [review]

File: sdk/log/internal/x/x.go (hunk 4 above)
Comment (pellared) [reply sdk/log/internal/x/x.go:46]: https://github.com/open-telemetry/opentelemetry-go/issues/5810 created

**Comment by pellared** [comment]
@dashpole, @dmathieu, can you please take another look as there have been some API changes (related to https://github.com/open-telemetry/opentelemetry-go/pull/5791#discussion_r1755013187)?

**Review by dashpole — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]
